src/ddmd/dinterpret.d

```diff
@@ -1050,6 +1164,48 @@ extern (C++) Expression interpret(FuncDeclaration fd, InterState* istate, Expres
         e = CTFEExp.cantexp;
     }
 
+    if (global.params.cache && globalExpr)
+    {
+        clock_t ctfeEnd = clock();
+        double delta = 1.0*(ctfeEnd - ctfeStart) / CLOCKS_PER_SEC;
+        if (delta > 0.1)
+        {
+            debug(DCACHE)
+            {
+                printf("interpreting global %s: %s\n", fd.loc.toChars(), fd.toFullSignature());
+                foreach (i, earg; eargs[])
+                {
+                    printf("\targ %d : %s\n", i, earg.toChars());
+                }
+            }
+            OutBuffer buf;
+            scope extractor = new ExtractSymbols(fd._scope);
+            scope exprVisitor = new PostorderExpressionVisitor(extractor);
+            e.accept(exprVisitor);
+            extern(C) static int comp(scope const(void*) *a , scope const(void*)* b)
+            {
+                return a > b ? 1 : (a == b ? 0 : -1);
+            }
+            qsort(extractor.symbols.ptr, extractor.symbols.length, Dsymbol.sizeof, cast(_compare_fp_t)&comp);
+            uniqueInPlace(extractor.symbols);
+            buf.writestring("(){\n");
+            foreach (t; extractor.symbols)
+            {
+                buf.writestring("import ");
+                writeModuleName(t, buf);
+                buf.writestring(" : ");
+                buf.writestring(t.toChars());
+                buf.writestring(";\n");
+            }
+            buf.writestring("return ");
+            buf.writestring(e.toChars());
```

Comment:
`toChars` was never intended as parseable representation of AST.
If you want to serialize sth. you should look at hdrgen or might even need to implement sth. based on that.

Reply:
Or serialize the the AST so the output is binary-like. Do I understand that this writes out plain text to cache? What's there to prevent tampering?

Reply:
Well we're not really in the business of encrypting intermediate formats so no worries there.